_posts/2024-08-06-quarkus-and-leyden.adoc

```diff
@@ -0,0 +1,275 @@
+No other runtime is able to match the JVM on this.
+
+As we all know, a Java runtime does not directly run Java source code. The content of our JAR file is not executable machine code, but Java bytecode generated from Java source code using the javac compiler.
+
+A key feature of bytecode is portability, encoding the structure of Java classes and operation of their methods in a machine and operating-system independent format. A Java runtime obeys the type information in the bytecode when laying out Java objects. Execution of a method normally involves interpreting the operations in the method bytecode, although a runtime may also choose to compile method bytecode to equivalent, native machine code and execute the latter directly. 
+
+The unit of delivery for bytecode is a class file, which models a single class. The Java runtime itself provides a host of utility and runtime management classes, as class files embedded in either system jars or jmod files. Applications supplement this with their own class files, usually by appending jars to the classpath or module path.
+
+Bytecode is delivered class-at-a-time to allow the runtime to load classes _lazily_: i.e. the runtime will only lookup, verify and consume a class file when that class's definition is required to proceed with execution.
+
+Lazy loading is what allows Java to be a dynamic language -- i.e. one where the code that is included in the program can be decided at runtime. That might include loading classes from jars identified at runtime, possibly loaded via the network. Alternatively, it might include generating class bytecode at runtime, as is done with proxy classes or service provider auxiliary classes.
+
+=== Just in Time (JIT) and Ahead of Time (AOT)
+
+Another name to describe Java's lazy loading is 'Just in Time (JIT). JIT is a well known term used to describe the operation of Java's runtime compilers. What is less well known is that it has a much wider use. Many other operations performed by the JVM are done lazily at runtime or 'Just In Time'.
+
+An alternative to doing things 'Just in Time' (JIT) is to do them 'Ahead Of Time' (AOT). For example, Graal's Native Image runtime loads and analyses the bytecode of every single class needed by an application, including JDK runtime classes, 'Ahead Of Time' i.e. at image build time. It uses the type and method information encoded in that bytecode to 'Ahead Of Time' compile a complete program that includes code for every method that might possibly be executed by the application. 
+
+Graal Native Image lies at one extreme, everything is done AOT, and the traditional Java runtime model lies at the other extreme, as much as possible is done JIT. However, it is actually possible to mix and match AOT and JIT models of execution in one runtime: re-balancing that AOT vs JIT mix is the goal of the first EA release of project Leyden.
```

Comment:
when mix/match of aot and jit could we relate/compare this to graalvm's profile guided optimization?

Reply:
We could talk about PGO (not just PGO in GraalVM) as there are similarities. However, there are also major differences with other ways of doing PGO. I don't think we want to open this can of worms as it will derail the flow of the post from where we really need to go.

Reply:
Right we don't want to go too much on a tangent with PGO but I'll add a short reference to it to be fair to GraalVM, as it does relate conceptually to this even if it's not the same.

Reply:
I'll be adding this paragraph:
```
A native image build might also take advantage of Profile Guided Optimisations (PGO), which allows it to leverage some data about what's presumably happening at runtime back into the compilation process, guiding its optimisations. However, it's only peeking into a simulation of runtime metrics, and ultimately, the compiler still needs to make all optimisation tradeoffs Ahead Of Time; this has pros and cons. The primary disadvantage is that any suboptimal or outright bad decision is cast in stone. The advantage is more decisive for short-lived applications as the tradeoff of carrying all support for JIT optimisations in the runtime is less justifiable when there is barely an opportunity to take advantage of it.
```